internal/database/repo_kvps.go

```
@@ -49,14 +48,19 @@ func (s *repoKVPStore) Create(ctx context.Context, repoID api.RepoID, kvp KeyVal
 	q := `
 	INSERT INTO repo_kvps (repo_id, key, value)
 	VALUES (%s, %s, %s)
+	ON CONFLICT DO NOTHING
 	`
 
-	if err := s.Exec(ctx, sqlf.Sprintf(q, repoID, kvp.Key, kvp.Value)); err != nil {
-		var e *pgconn.PgError
-		if errors.As(err, &e) && e.Code == "23505" {
-			return errors.Newf(`Repo metadata key "%s" already exists for the given repository.`, kvp.Key)
-		}
-		return err
+	result, err := s.ExecResult(ctx, sqlf.Sprintf(q, repoID, kvp.Key, kvp.Value))
+	if err != nil {
+		return errors.Wrap(err, "executing SQL query")
+	}
+	nrows, err := result.RowsAffected()
+	if err != nil {
+		return errors.Wrap(err, "getting rows affected")
+	}
+	if nrows == 0 {
```

Sorry, but I actually misled you about the number of rows affected.
The thing is, it is impossible to understand whether the conflict has occurred (i.e. the entry already existed) with `ON CONFLICT DO NOTHING` clause. 

The SQL would get a bit messy if we want to distinguish between `INSERT 0 1` response from Postgres in case of _inserting an entry_ or _not inserting a duplicate.

If we really need this error (instead of silently skipping the duplicate), then we can use 2 separate queries, first to check existence and return an error if exists, and second to insert the entry which we (now) know, has no existing duplicates.

If we don't need this error, then you can just skip the 58-63 lines altogether.

Anyway, it's your call, sorry for confusion!

---

+1 on that. Also, if we want to do 2 separate queries, we need to do them in transaction to keep the DB consistency in mind.

---

We could also keep the SQL statement as it was and just catch the error properly here. E.g. if there's a conflict, we should get a specific error from postgres that we could react to.